orttraining/orttraining/python/training/ortmodule/__init__.py

```diff
@@ -17,33 +23,35 @@
 MINIMUM_RUNTIME_PYTORCH_VERSION_STR = '1.8.1'
 TORCH_CPP_DIR = os.path.join(os.path.dirname(__file__),
                              'torch_cpp_extensions')
+_FALLBACK_INIT_EXCEPTION = None
+ORTMODULE_FALLBACK_POLICY = _FallbackPolicy.FALLBACK_DISABLE
+ORTMODULE_FALLBACK_RETRY = True
 
 # Verify minimum PyTorch version is installed before proceding to ONNX Runtime initialization
 try:
     import torch
     runtime_pytorch_version = version.parse(torch.__version__.split('+')[0])
     minimum_runtime_pytorch_version = version.parse(MINIMUM_RUNTIME_PYTORCH_VERSION_STR)
     if runtime_pytorch_version < minimum_runtime_pytorch_version:
-        raise RuntimeError(
-            f'ONNX Runtime ORTModule frontend requires PyTorch version greater or equal to {MINIMUM_RUNTIME_PYTORCH_VERSION_STR}, '
-            f'but version {torch.__version__} was found instead.')
-except:
-    raise RuntimeError(f'PyTorch {MINIMUM_RUNTIME_PYTORCH_VERSION_STR} must be installed in order to run ONNX Runtime ORTModule frontend!')
+        raise wrap_exception(ORTModuleInitException,
+                             RuntimeError(
+                                 f'ONNX Runtime ORTModule frontend requires PyTorch version greater or equal to {MINIMUM_RUNTIME_PYTORCH_VERSION_STR}, '
+                                 f'but version {torch.__version__} was found instead.'))
+except ORTModuleFallbackException as e:
+    # Initialization fallback is handled at ORTModule.__init__
+    _FALLBACK_INIT_EXCEPTION = e
+except ImportError as e:
+    raise RuntimeError(f'PyTorch {MINIMUM_RUNTIME_PYTORCH_VERSION_STR} must be installed in order to run ONNX Runtime ORTModule frontend!') from e
 
 # Verify whether PyTorch C++ extensions are already compiled
-torch_cpp_exts = glob(os.path.join(TORCH_CPP_DIR, '*.so'))
-torch_cpp_exts.extend(glob(os.path.join(TORCH_CPP_DIR, '*.dll')))
-torch_cpp_exts.extend(glob(os.path.join(TORCH_CPP_DIR, '*.dylib')))
-if not torch_cpp_exts and '-m' not in sys.argv:
-    raise EnvironmentError(f"ORTModule's extensions were not detected at '{TORCH_CPP_DIR}' folder. "
-                           "Run `python -m torch_ort.configure` before using `ORTModule` frontend.")
 
-# PyTorch custom Autograd function support
-from ._custom_autograd_function import enable_custom_autograd_support
-enable_custom_autograd_support()
+if not is_torch_cpp_extensions_installed(TORCH_CPP_DIR) and '-m' not in sys.argv:
```

Comment:
This is existing code related to torch cpp extensions, not fallback.

But the reasoning behind this is that when the user calls `python -m torch_ort.configure` or `python -m onnxruntime.training.ortmodule.torch_cpp_extensions.install`, we detect the `-m` and compile the extension as opposed to raise an exception complaining about the non compiled extensions.

We need this because to reach the `install.py` inside `onnxruntime`, onxruntime import must succeed.